packages/mds-compliance/api.ts

```diff
@@ -143,40 +136,38 @@ function api(app: express.Express): express.Express {
           const filteredEvents = compliance_engine.filterEvents(events)
           const result = compliance_engine.processPolicy(policy, filteredEvents, geographies, deviceMap)
           if (result === undefined) {
-            return res.status(400).send({ err: 'bad_param' })
+            return res.status(400).send(new BadParamsError('Unable to process compliance results'))
           }
-          return res.status(200).send(result)
+          return res.status(200).send({ ...result, timestamp: query_date })
         }
       } else {
-        return res.status(401).send({ err: 'Unauthorized' })
+        return res.status(401).send(new AuthorizationError())
       }
     } catch (err) {
-      if (err.message.includes('not_found')) {
-        return res.status(400).send({ err: 'bad_param' })
-      }
-      await fail(err)
+      return res.status(500).send(new ServerError())
     }
   })
 
   app.get(pathsFor('/count/:rule_id'), async (req: ComplianceApiRequest, res: ComplianceApiResponse) => {
-    if (!AllowedProviderIDs.includes(res.locals.provider_id)) {
-      return res.status(401).send({ result: 'unauthorized access' })
+    const { timestamp } = {
+      ...parseRequest(req, { parser: Number }).query('timestamp')
     }
-
-    async function fail(err: Error) {
-      logger.error(err.stack || err)
-      if (err.message.includes('invalid rule_id')) {
-        return res.status(404).send(err.message)
-      }
-      /* istanbul ignore next */
-      return res
-        .status(500)
-        .send({ error: 'server_error', error_description: 'an internal server error has occurred and been logged' })
+    const query_date = timestamp || now()
+    if (!AllowedProviderIDs.includes(res.locals.provider_id)) {
+      return res.status(401).send(new AuthorizationError('Unauthorized'))
     }
 
     const { rule_id } = req.params
     try {
-      const rule = await db.readRule(rule_id)
+      const policies = await db.readActivePolicies(query_date)
+      const filteredPolicies = policies.filter(policy => {
```

Comment:
Instead of doing this filtering in TS, we could adapt the SQL query to look for a matching `rule_id` in the policy json.

Reply:
Ought to also just be destructuring to pull off the head like `const [policy] = db.readActivePolicies(...)`, I see that on line 209 the policy that we send in the response is not necessarily the policy we considered when computing compliance

Reply:
> Ought to also just be destructuring to pull off the head like `const [policy] = db.readActivePolicies(...)`, I see that on line 209 the policy that we send in the response is not necessarily the policy we considered when computing compliance

I don't see how that's possible, now that rule uniqueness is actually enforced.

Reply:
Because `filteredPolicies !== policies`. We consider the head of the filteredPolicies for the evaluation, but in the response body we are returning `policies[0]`.

Reply:
So the policy that we return in the payload may not be a policy that even has the `rule_id` in question